### ISSUE TITLE
Arrow2: db-benchmark q9 fails with indexs out of bounds

### ISSUE THREAD

**Issue by ritchie46** [pola-rs/polars#887]
Only occurs on dataset with null values. Also only occurs at a certain dataset size, so needs further investigation.

# Reproduce

```python
x = pl.read_csv("G1_1e7_1e2_5_0.csv", 
    dtype={"id4":pl.Int32, "id5":pl.Int32, "id6":pl.Int32, "v1":pl.Int32, "v2":pl.Int32, "v3":pl.Float64},
    stop_after_n_rows=100000)

x["id1"] = x["id1"].cast(pl.Categorical)
x["id2"] = x["id2"].cast(pl.Categorical)
x["id3"] = x["id3"].cast(pl.Categorical)

(x.lazy().groupby(["id2","id4"])
    .agg((pl.pearson_corr("v1","v2")**2).alias("r2"))
    .collect())
```

# Stacktrace
```
thread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/bitmap/utils/chunk_iterator/mod.rs:156:27
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/panicking.rs:92:14
   2: core::panicking::panic_bounds_check
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/panicking.rs:69:5
   3: <arrow2::bitmap::utils::chunk_iterator::BitChunks<T> as core::iter::traits::iterator::Iterator>::next
             at /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/bitmap/utils/chunk_iterator/mod.rs:156:27
   4: <&mut I as core::iter::traits::iterator::Iterator>::next
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/traits/iterator.rs:3443:9
   5: core::iter::traits::iterator::Iterator::fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/traits/iterator.rs:2111:29
   6: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:122:9
   7: <usize as core::iter::traits::accum::Sum>::sum
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/traits/accum.rs:44:17
   8: core::iter::traits::iterator::Iterator::sum
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/traits/iterator.rs:2966:9
   9: arrow2::bitmap::utils::null_count
             at /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/bitmap/utils/mod.rs:71:28
  10: arrow2::bitmap::immutable::Bitmap::slice
             at /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/bitmap/immutable.rs:111:27
  11: arrow2::array::primitive::PrimitiveArray<T>::slice::{{closure}}
             at /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/array/primitive/mod.rs:79:54
  12: core::option::Option<T>::map
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/option.rs:489:29
  13: arrow2::array::primitive::PrimitiveArray<T>::slice
             at /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/array/primitive/mod.rs:79:24
  14: <arrow2::array::primitive::PrimitiveArray<T> as arrow2::array::Array>::slice
             at /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/array/primitive/mod.rs:154:18
  15: arrow2::array::list::ListArray<O>::value
             at /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/array/list/mod.rs:83:13
  16: arrow2::array::list::iterator::<impl arrow2::array::IterableListArray for arrow2::array::list::ListArray<O>>::value
             at /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/array/list/iterator.rs:60:9
  17: <arrow2::array::list::iterator::ListValuesIter<A> as core::iter::traits::iterator::Iterator>::next
             at /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/array/list/iterator.rs:35:14
  18: <arrow2::bitmap::utils::zip_validity::ZipValidity<T,I> as core::iter::traits::iterator::Iterator>::next
             at /home/ritchie46/.cargo/git/checkouts/arrow2-8a2ad61d97265680/4eed0d0/src/bitmap/utils/zip_validity.rs:38:25
  19: <core::iter::adapters::flatten::FlattenCompat<I,U> as core::iter::traits::iterator::Iterator>::next
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/flatten.rs:262:23
  20: <core::iter::adapters::flatten::Flatten<I> as core::iter::traits::iterator::Iterator>::next
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/flatten.rs:168:9
  21: <polars_arrow::utils::TrustMyLength<I,J> as core::iter::traits::iterator::Iterator>::next
             at /home/ritchie46/code/polars/polars/polars-arrow/src/utils.rs:27:9
  22: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:101:9
  23: <alloc::boxed::Box<I,A> as core::iter::traits::iterator::Iterator>::next
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/alloc/src/boxed.rs:1509:9
  24: <core::iter::adapters::zip::Zip<A,B> as core::iter::adapters::zip::ZipImpl<A,B>>::next
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/zip.rs:151:17
  25: <core::iter::adapters::zip::Zip<A,B> as core::iter::traits::iterator::Iterator>::next
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/zip.rs:77:9
  26: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:101:9
  27: polars_core::chunked_array::upstream_traits::<impl core::iter::traits::collect::FromIterator<core::option::Option<Ptr>> for polars_core::chunked_array::ChunkedArray<polars_core::datatypes::ListType>>::from_iter
             at /home/ritchie46/code/polars/polars/polars-core/src/chunked_array/upstream_traits.rs:194:22
  28: core::iter::traits::iterator::Iterator::collect
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/traits/iterator.rs:1748:9
  29: <polars_lazy::physical_plan::expressions::binary_function::BinaryFunctionExpr as polars_lazy::physical_plan::expressions::PhysicalAggregation>::aggregate
             at /home/ritchie46/code/polars/polars/polars-lazy/src/physical_plan/expressions/binary_function.rs:81:18
  30: <polars_lazy::physical_plan::expressions::alias::AliasExpr as polars_lazy::physical_plan::expressions::PhysicalAggregation>::aggregate
             at /home/ritchie46/code/polars/polars/polars-lazy/src/physical_plan/expressions/alias.rs:57:23
  31: <polars_lazy::physical_plan::expressions::apply::ApplyExpr as polars_lazy::physical_plan::expressions::PhysicalAggregation>::aggregate::{{closure}}
             at /home/ritchie46/code/polars/polars/polars-lazy/src/physical_plan/expressions/apply.rs:70:25
  32: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &F>::call_mut
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/ops/function.rs:247:13
  33: core::iter::adapters::map::map_try_fold::{{closure}}
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:89:28
  34: core::iter::traits::iterator::Iterator::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/traits/iterator.rs:1972:21
  35: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:115:9
  36: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:115:9
  37: <core::iter::adapters::take_while::TakeWhile<I,P> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/take_while.rs:93:13
  38: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:115:9
  39: <core::iter::adapters::take_while::TakeWhile<I,P> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/take_while.rs:93:13
  40: <core::iter::adapters::take_while::TakeWhile<I,P> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/take_while.rs:108:9
  41: <rayon::iter::fold::FoldFolder<C,ID,F> as rayon::iter::plumbing::Folder<T>>::consume_iter
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/fold.rs:158:20
  42: <rayon::iter::while_some::WhileSomeFolder<C> as rayon::iter::plumbing::Folder<core::option::Option<T>>>::consume_iter
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/while_some.rs:139:21
  43: <rayon::iter::map::MapFolder<C,F> as rayon::iter::plumbing::Folder<T>>::consume_iter
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/map.rs:248:21
  44: <rayon::iter::map::MapFolder<C,F> as rayon::iter::plumbing::Folder<T>>::consume_iter
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/map.rs:248:21
  45: rayon::iter::plumbing::Producer::fold_with
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/plumbing/mod.rs:110:9
  46: rayon::iter::plumbing::bridge_producer_consumer::helper
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/plumbing/mod.rs:438:13
  47: rayon::iter::plumbing::bridge_producer_consumer
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/plumbing/mod.rs:397:12
  48: <rayon::iter::plumbing::bridge::Callback<C> as rayon::iter::plumbing::ProducerCallback<I>>::callback
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/plumbing/mod.rs:373:13
  49: <rayon::slice::Iter<T> as rayon::iter::IndexedParallelIterator>::with_producer
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/slice/mod.rs:525:9
  50: rayon::iter::plumbing::bridge
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/plumbing/mod.rs:357:12
  51: <rayon::slice::Iter<T> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/slice/mod.rs:501:9
  52: <rayon::iter::map::Map<I,F> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/map.rs:49:9
  53: <rayon::iter::map::Map<I,F> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/map.rs:49:9
  54: <rayon::iter::while_some::WhileSome<I> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/while_some.rs:44:9
  55: <rayon::iter::fold::Fold<I,ID,F> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/fold.rs:59:9
  56: <rayon::iter::map::Map<I,F> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/map.rs:49:9
  57: rayon::iter::reduce::reduce
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/reduce.rs:15:5
  58: rayon::iter::ParallelIterator::reduce
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/mod.rs:981:9
  59: rayon::iter::extend::collect
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/extend.rs:29:5
  60: rayon::iter::collect::<impl rayon::iter::ParallelExtend<T> for alloc::vec::Vec<T>>::par_extend
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/collect/mod.rs:165:28
  61: rayon::iter::from_par_iter::collect_extended
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/from_par_iter.rs:17:5
  62: rayon::iter::from_par_iter::<impl rayon::iter::FromParallelIterator<T> for alloc::vec::Vec<T>>::from_par_iter
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/from_par_iter.rs:30:9
  63: rayon::iter::ParallelIterator::collect
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/mod.rs:2046:9
  64: rayon::result::<impl rayon::iter::FromParallelIterator<core::result::Result<T,E>> for core::result::Result<C,E>>::from_par_iter
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/result.rs:121:26
  65: rayon::iter::ParallelIterator::collect
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/mod.rs:2046:9
  66: <polars_lazy::physical_plan::expressions::apply::ApplyExpr as polars_lazy::physical_plan::expressions::PhysicalAggregation>::aggregate
             at /home/ritchie46/code/polars/polars/polars-lazy/src/physical_plan/expressions/apply.rs:65:30
  67: <polars_lazy::physical_plan::expressions::alias::AliasExpr as polars_lazy::physical_plan::expressions::PhysicalAggregation>::aggregate
             at /home/ritchie46/code/polars/polars/polars-lazy/src/physical_plan/expressions/alias.rs:57:23
  68: polars_lazy::physical_plan::executors::groupby::groupby_helper::{{closure}}::{{closure}}::{{closure}}
             at /home/ritchie46/code/polars/polars/polars-lazy/src/physical_plan/executors/groupby.rs:53:31
  69: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &F>::call_mut
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/ops/function.rs:247:13
  70: core::iter::adapters::map::map_try_fold::{{closure}}
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:89:28
  71: core::iter::traits::iterator::Iterator::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/traits/iterator.rs:1972:21
  72: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:115:9
  73: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:115:9
  74: <core::iter::adapters::take_while::TakeWhile<I,P> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/take_while.rs:93:13
  75: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/map.rs:115:9
  76: <core::iter::adapters::take_while::TakeWhile<I,P> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/take_while.rs:93:13
  77: <core::iter::adapters::take_while::TakeWhile<I,P> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/f586d79d183d144e0cbf519e29247f36670e2076/library/core/src/iter/adapters/take_while.rs:108:9
  78: <rayon::iter::fold::FoldFolder<C,ID,F> as rayon::iter::plumbing::Folder<T>>::consume_iter
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/fold.rs:158:20
  79: <rayon::iter::while_some::WhileSomeFolder<C> as rayon::iter::plumbing::Folder<core::option::Option<T>>>::consume_iter
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/while_some.rs:139:21
  80: <rayon::iter::map::MapFolder<C,F> as rayon::iter::plumbing::Folder<T>>::consume_iter
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/map.rs:248:21
  81: <rayon::iter::map::MapFolder<C,F> as rayon::iter::plumbing::Folder<T>>::consume_iter
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/map.rs:248:21
  82: rayon::iter::plumbing::Producer::fold_with
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/plumbing/mod.rs:110:9
  83: rayon::iter::plumbing::bridge_producer_consumer::helper
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/plumbing/mod.rs:438:13
  84: rayon::iter::plumbing::bridge_producer_consumer
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/plumbing/mod.rs:397:12
  85: <rayon::iter::plumbing::bridge::Callback<C> as rayon::iter::plumbing::ProducerCallback<I>>::callback
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/plumbing/mod.rs:373:13
  86: <rayon::slice::Iter<T> as rayon::iter::IndexedParallelIterator>::with_producer
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/slice/mod.rs:525:9
  87: rayon::iter::plumbing::bridge
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/plumbing/mod.rs:357:12
  88: <rayon::slice::Iter<T> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/slice/mod.rs:501:9
  89: <rayon::iter::map::Map<I,F> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/map.rs:49:9
  90: <rayon::iter::map::Map<I,F> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/map.rs:49:9
  91: <rayon::iter::while_some::WhileSome<I> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/while_some.rs:44:9
  92: <rayon::iter::fold::Fold<I,ID,F> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/ritchie46/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-1.5.1/src/iter/fold.rs:59:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

```

**Comment by jorgecarleitao** [comment]
This is arrow2's fault and has been fixed in main 👍

**Comment by ritchie46** [comment]
That means we are there. :100: :partying_face:

**Comment by jorgecarleitao** [comment]
is there any changes to the benchmark results? :P

**Comment by ritchie46** [comment]
Fixed in #895

benchmarks not yet representative. WIP.
